File: atvi/modelos/cliente.ts
import Prototipo from "../interfaces/prototipo"
import Documento from "./documento"
import Endereco from "./endereco"
import Telefone from "./telefone"

export default class Cliente implements Prototipo {
    public nome: string
    public nomeSocial: string
    public dataNascimento: Date
    public dataCadastro: Date
    public telefones: Telefone[] = []
    public endereco: Endereco
    public documentos: Documento[] = []
    public dependentes: Cliente[] = []
    public titular: Cliente

    public clonar(): Prototipo {
        const clone = new Cliente()
        clone.nome = this.nome
        clone.nomeSocial = this.nomeSocial
        clone.dataNascimento = this.dataNascimento
        clone.dataCadastro = this.dataCadastro
        clone.telefones = this.telefones.map(telefone => telefone.clonar() as Telefone)
        clone.endereco = this.endereco.clonar() as Endereco
        clone.documentos = this.documentos.map(documento => documento.clonar() as Documento)
        clone.dependentes = this.dependentes.map(dependente => dependente.clonar() as Cliente)
        clone.titular = this.titular ? this.titular.clonar() as Cliente : null
        return clone
    }
}

File: AtvV/BackEnd/src/main/java/com/joao/AtvV/models/Cliente.java
package com.joao.AtvV.models;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Cliente {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    private String nome;

    private LocalDate dataNascimento;

    private LocalDateTime dataCadastro;

    @OneToMany(mappedBy = "cliente", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Documentos> documentos;

    @OneToMany(mappedBy = "cliente", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Telefone> telefones;

    @OneToMany(mappedBy = "cliente", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Endereco> enderecos;
}
